File: Leetcode 3901-4000/3938/3938.java
// Leetcode 3938: Maximum Path Intersection Sum in a Grid
// https://leetcode.com/problems/maximum-path-intersection-sum-in-a-grid/
// Solved on 7th of June, 2026
class Solution {
    /**
     * Calculates the maximum path intersection sum in a grid.
     * 
     * @param grid A 2D integer array representing the grid.
     * @return The maximum score calculated based on path intersections.
     */
    public int maxScore(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int ans = Integer.MIN_VALUE;

        for (int r = 0; r < m; r++) {
            int max1 = grid[r][0];
            for (int c = 1; c < n; c++) {
                int max2 = max1 + grid[r][c];
                if (max2 > ans) {
                    ans = max2;
                }
                max1 = Math.max(grid[r][c], max1 + grid[r][c]);
            }
        }

        for (int c = 0; c < n; c++) {
            int max1 = grid[0][c];
            for (int r = 1; r < m; r++) {
                int max2 = max1 + grid[r][c];
                if (max2 > ans) {
                    ans = max2;
                }
                max1 = Math.max(grid[r][c], max1 + grid[r][c]);
            }
        }

        for (int r = 1; r < m - 1; r++) {
            for (int c = 1; c < n - 1; c++) {
                if (grid[r][c] > ans) {
                    ans = grid[r][c];
                }
            }
        }

        return ans;
    }
}

File: Leetcode 3901-4000/3938/3938.py
# Leetcode 3938: Maximum Path Intersection Sum in a Grid
# https://leetcode.com/problems/maximum-path-intersection-sum-in-a-grid/
# Solved on 7th of June, 2026
class Solution:
    def maxScore(self, grid: list[list[int]]) -> int:
        """
        Calculates the maximum path intersection sum in a given grid.

        :param grid: A 2D list of integers representing the grid.
        :return: An integer representing the maximum score calculated from path intersections.
        """
        numRows = len(grid)
        numCols = len(grid[0])
        maxSum = -float('inf')

        for rowIdx in range(numRows):
            if numCols >= 2:
                prevDp = grid[rowIdx][0] + grid[rowIdx][1]
                if prevDp > maxSum:
                    maxSum = prevDp
                prevVal = grid[rowIdx][1]

                for colIdx in range(2, numCols):
                    curDp = grid[rowIdx][colIdx] + (prevDp if prevDp > prevVal else prevVal)
                    if curDp > maxSum:
                        maxSum = curDp

                    prevDp = curDp
                    prevVal = grid[rowIdx][colIdx]

        for colIdx in range(numCols):
            if numRows >= 2:
                prevDp = grid[0][colIdx] + grid[1][colIdx]
                if prevDp > maxSum:
                    maxSum = prevDp
                prevVal = grid[1][colIdx]

                for rowIdx in range(2, numRows):
                    curDp = grid[rowIdx][colIdx] + (prevDp if prevDp > prevVal else prevVal)
                    if curDp > maxSum:
                        maxSum = curDp

                    prevDp = curDp
                    prevVal = grid[rowIdx][colIdx]

        for rowIdx in range(1, numRows - 1):
            for colIdx in range(1, numCols - 1):
                if grid[rowIdx][colIdx] > maxSum:
                    maxSum = grid[rowIdx][colIdx]

        return maxSum